lib/c-api/include/clingo/config.h

```diff
@@ -42,9 +42,92 @@ enum clingo_config_type_e {
+//! in the path. It is up to the entry implementation to interpret an absent
+//! index (i.e., when `index` is NULL).
+//!
+//! All callbacks are optional; if a callback is NULL, the corresponding
+//! operation is not supported for this entry.
+typedef struct clingo_config_entry {
+    //! Query value flags for this entry.
+    //!
+    //! @param[in] index an optional array index (can be NULL)
+    //! @param[in] data user data pointer
+    //! @param[out] info output flags (see ::clingo_config_value_flags)
+    //! @return true on success, false on error
+    bool (*info)(size_t *index, void *data, clingo_config_value_flags_t *info);
+
+    //! Get the value of this entry as a string.
+    //!
+    //! @param[in] index an optional array index (can be NULL)
+    //! @param[in] data user data pointer
+    //! @param[out] value output string value
+    //! @param[out] has_value set to true if a value is available, false otherwise
+    //! @return true on success, false on error
+    bool (*get)(size_t *index, void *data, clingo_string_t *value, bool *has_value);
+
+    //! Set the value of this entry from a string.
+    //!
+    //! @param[in] index an optional array index (can be NULL)
+    //! @param[in] value input string value
+    //! @param[in] size length of the value string
+    //! @param[in] data user data pointer
+    //! @return true on success, false on error
+    bool (*set)(size_t *index, char const *value, size_t size, void *data);
+
+    //! Get the size of this entry if it is an array.
+    //!
+    //! @param[in] data user data pointer
+    //! @param[out] size number of elements in the array (or zero)
+    //! @param[out] has_size true if the entry is an array, false otherwise
+    //! @return true on success, false on error
+    bool (*size)(void *data, size_t *size, bool *has_size);
+
+    //! Free the user data associated with this entry.
+    //!
+    //! @param[in] data user data pointer
+    void (*free)(void *data);
+} clingo_config_entry_t;
+
+//! Add a custom configuration entry under a parent key.
+//!
+//! The entry is defined by the provided callbacks and user data. The API takes
+//! ownership of the `data` pointer and will call the `free` callback (if non-NULL)
+//! when the entry is removed or the configuration is destroyed.
+//!
+//! @param[in] config the configuration object
+//! @param[in] parent the parent key under which to add the entry
+//! @param[in] name the name of the new entry (not null-terminated)
+//! @param[in] name_size the length of the name string
+//! @param[in] description the description of the entry (not null-terminated)
+//! @param[in] description_size the length of the description string
+//! @param[in] entry pointer to the entry callbacks struct
+//! @param[in] data user data pointer for the entry callbacks
+//! @return true on success, false on error
+CLINGO_VISIBILITY_DEFAULT bool clingo_config_add(clingo_config_t *config, clingo_id_t parent, char const *name,
```

Comment:
It is consistent with other API functions and you probably already explained this to me but could you briefly reiterate why we pass "string views" as two parameters but receive them as a `clingo_string_t`?

Reply:
Also doing the input would prob be good. Its just a pain to change.

Reply:
> Also doing the input would prob be good. Its just a pain to change.

...but you have AI now to alleviate the pain :sweat_smile: